pkg/strategy/grid2/strategy.go

```diff
@@ -1385,9 +1385,36 @@ func (s *Strategy) recoverGridWithOpenOrdersByScanningTrades(ctx context.Context
 	}
 
 	// 1. build pin-order map
-	// 2. fill the pin-order map by querying trades
+	pinOrderMapWithOpenOrders, err := s.buildPinOrderMap(grid, openOrdersOnGrid)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build pin order map with open orders")
+	}
+
+	// 2. build the filled pin-order map by querying trades
+	pinOrderMapWithFilledOrders, err := s.buildFilledPinOrderMapFromTrades(ctx, historyService, pinOrderMapWithOpenOrders)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build filled pin order map")
+	}
+
 	// 3. get the filled orders from pin-order map
+	filledOrders := getOrdersFromPinOrderMapInAscOrder(pinOrderMapWithFilledOrders)
+	numsFilledOrders := len(filledOrders)
+	i := 0
+	if numsFilledOrders == int(expectedOrderNums-openOrdersOnGridNums) {
+		// nums of filled order is the same as Size - 1 - num(open orders)
+	} else if numsFilledOrders == int(expectedOrderNums-openOrdersOnGridNums+1) {
+		i++
+	} else {
+		return fmt.Errorf("not reasonable num of filled orders")
+	}
+
 	// 4. emit the filled orders
+	activeOrderBook := s.orderExecutor.ActiveMakerOrders()
+	for ; i < numsFilledOrders; i++ {
+		filledOrder := filledOrders[i]
+		s.logger.Infof("[DEBUG] emit filled order: %s (%s)", filledOrder.String(), filledOrder.UpdateTime)
+		activeOrderBook.EmitFilled(filledOrder)
```

Comment:
might not be easy to write tests for this?

Reply:
Use this flag to find my log easily XD

Reply:
I think it's not hard, we can OnFilled another assertion function.